src/bgw/job.c

```diff
@@ -723,11 +723,23 @@ get_job_lock_for_delete(int32 job_id)
 			proc = BackendIdGetProc(vxid->backendId);
 			if (proc != NULL && proc->isBackgroundWorker)
 			{
-				elog(NOTICE,
-					 "cancelling the background worker for job %d (pid %d)",
-					 job_id,
-					 proc->pid);
-				DirectFunctionCall1(pg_cancel_backend, Int32GetDatum(proc->pid));
+				/* Simply assuming that this pid corresponds to the background worker
+				 * running the job is not sufficient. The scheduler could also be the
+				 * one holding the lock, when transitioning the state of the job back
+				 * to scheduled state. So we must check we don't kill the scheduler.
+				 * See https://github.com/timescale/timescaledb/issues/5224
+				 */
+				const char *worker_name = GetBackgroundWorkerTypeByPid(proc->pid);
+				const char *bgw_scheduler_name = "TimescaleDB Background Worker Scheduler";
```

Comment:
What is the reason for defining `bgw_scheduler_name` here? Couldn't we use [this define](https://github.com/timescale/timescaledb/blob/d9849325d0d0f81a13db1e41aa56f8b567945e72/src/bgw/scheduler.c#L44)?

Reply:
Thank you, fixed